File: Array (MS)/347_Top_K_Frequent_Elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  let map = new Map();
  nums.forEach((num) => {
    map.has(num) ? map.set(num, map.get(num) + 1) : map.set(num, 1);
  });

  const sortedMap = new Map([...map.entries()].sort((a, b) => b[1] - a[1]));
  const iterator = sortedMap.entries();
  const result = [];

  for (let i = 0; i < k; i++) {
    result.push(iterator.next().value[0]);
  }

  return result;
};

// Runtime - 72ms   | Beats 77.22%of users with JavaScript
// Memory - 44.77mb | Beats 79.76%of users with JavaScript

File: Array (MS)/347_Top_K_Frequent_Elements.py
class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        record = {}

        for num in nums:
            if num in record:
                record[num] = record[num] + 1
            else:
                record[num] = 1

        sorted_record = sorted(record.items(), key=lambda x:x[1])
        print(sorted_record)
        result = []

        for _ in range(k):
            result.append(sorted_record.pop()[0])

        return sorted(result)



if __name__ == "__main__":
    solution = Solution()
    print(solution.topKFrequent([1, 1, 2, 2, 2, 2, 3], 2))

